edward/criticisms/evaluate.py

```diff
@@ -155,6 +160,10 @@ def evaluate(metrics, data, n_samples=500, output_key=None):
       evaluations += [categorical_crossentropy(y_true, y_pred)]
     elif metric == 'sparse_categorical_crossentropy':
       evaluations += [sparse_categorical_crossentropy(y_true, y_pred)]
+    elif metric == 'kl_divergence':
+      y_true_ = y_true / total_count
```

Comment:
This works only if it went through the specific logic above, i.e., `output_key` is categorical, multinomial, or onehotcategorical. Can you extend this to raise an error if it doesn't follow these classes—or more ambitiously, to have it work for all random variables?

Reply:
An umbrella question-as-a-response, which will help guide my changes: 

Do we want users, for example, to be able to compute `categorical_crossentropy([0, 2, 3], [0, 1, 4])`?

Reply:
Ideally no, because cross entropy is a function of two inputs: labels and logits/probabilities. Do we know what it means if the second input are also labels?

(Actually, going through the code, the way we currently calculate `*_crossentropy` and `hinge` seems wrong for discrete variables. They should be calculated on the discrete variables' logits/probs parameter and not the discrete variable itself.)

Reply:
Why need cross-entropy be computed with labels and logits/probabilities? Isn't this purely a point of "mitigate-the-numerical-precision-issues" implementation?

Reply:
On my end, I'm thinking of the question on a higher level. Cross-entropy gives the (pedantic definition to follow — apologies...) "uncertainty inherent in using the probabilities in one distribution to predict the events of another." To my knowledge, I don't believe these probabilities need even be normalized/valid. As such, given two tensors with outermost-dimensions of equal size, I don't see any mechanistic problem in allowing users to compute this metric or others similar at their own risk.

To this effect, I could simply add the following between lines 117 and [118](https://github.com/blei-lab/edward/blob/35423f22feb1d2598decbfcef0c021260ed4882a/edward/criticisms/evaluate.py#L118):

```python
total_count = tf.cond(hasattr(output_key, 'total_count'), lambda: output_key.total_count, lambda: 1)
```

To the contrary, should you want to take a more opinionated stance, i.e. precluding certain metrics from being calculated given certain random-variables, I'd think to do this at the beginning of the function. A priori, we'd assign a list of valid metrics as a property on each `RandomVariable` class; in `evaluate`, we'd loop through `metrics` and check for conflicts.

Reply:
Got it. Unnormalized probabilities are okay. But the probabilities input need to have float type. That is, your example question tries to calculate cross entropy for two integer-valued label inputs.

Reply:
Why must they be floats? Isn't this a point of implementation as well?

I'll humbly interpret your comment to mean: as long as users pass inputs satisfying the `evaluate` API, they're free to evaluate with whatever metrics they want. Adding the snippet above between 117 and 118.

Reply:
That interpretation is reasonable. The only constraint is semantics: the second input _has_ to be a probability and not a label. With the current data types, discrete random variables with finite support and any `total_count` parameter must be integers and their probabilities must be floats.

Reply:
Got it. The floats imperative is Tensorflow-specific, yeah? This part of your comment confused me at first, but the more I read Tensorflow docs, the more I realize how much it prefers things to be floats.

Reply:
👀/thoughts on `compute_multinomial_mode`?

Additionally, edward-specific stylistic and comments-best-practices feedback appreciated!

Finally, I'm assuming I'll squash these commits together before a potential merge.

Reply:
Looks great so far!

No need to manually squash. Edward does a squash+merge when merging any PRs.

Reply:
Super. I'll finish up with "scikit-learn-style binary classification aggregation metrics" in the coming days.